pychromecast/discovery.py

```diff
@@ -108,28 +103,31 @@ def start_discovery(
     discovered chromecast's zeroconf name. This is the dictionary key to find
     the chromecast metadata in listener.services.
 
-    This method returns the CastListener object and the zeroconf ServiceBrowser
-    object. The CastListener object will contain information for the discovered
-    chromecasts. To stop discovery, call the stop_discovery method with the
-    ServiceBrowser object.
+    This method returns the zeroconf ServiceBrowser object.
+
+    A CastListener object must be passed, and will contain information for the
+    discovered chromecasts. To stop discovery, call the stop_discovery method with
+    the ServiceBrowser object.
 
     A shared zeroconf instance can be passed as zeroconf_instance. If no
     instance is passed, a new instance will be created.
     """
-    listener = CastListener(add_callback, remove_callback, update_callback)
     return (
-        listener,
         zeroconf.ServiceBrowser(
             zeroconf_instance or zeroconf.Zeroconf(),
             "_googlecast._tcp.local.",
             listener,
-        ),
+        )
     )
 
 
 def stop_discovery(browser):
     """Stop the chromecast discovery thread."""
-    browser.cancel()
+    try:
+        browser.cancel()
+    except RuntimeError:
+        # Throws if called from service callback when joining the zc browser thread
+        pass
```

Comment:
@bdraco `browser.cancel()` was added in #362. It throws if called from a service callback when trying to join. Not sure if the error should be ignored here, or if zeroconf should handle it?

Reply:
Since a thread can't wait for itself and `self.done` gets set to `True` trapping exception makes sense here.

zeroconf should probably be updated to skip the `.join` when `threading.current_thread()` is the `ServiceBrowser` thread

Reply:
We can use Thread Local data to check if we're in service browser data: https://docs.python.org/3.7/library/threading.html#thread-local-data